File: services/middleware-go/pagination.go
// Package middleware provides reusable pagination utilities for 54Bank Go microservices.
//
// Usage:
//
//	page := middleware.ParsePagination(r, 25) // default pageSize=25
//	items := middleware.ApplyPagination(allItems, page)
//	middleware.WritePaginatedJSON(w, items, len(allItems), page)
package middleware

import (
	"encoding/json"
	"net/http"
	"strconv"
)

// Pagination holds parsed pagination parameters.
type Pagination struct {
	Page     int `json:"page"`
	PageSize int `json:"pageSize"`
	Offset   int `json:"offset"`
}

// PaginatedResponse is the standard paginated API response envelope.
type PaginatedResponse struct {
	Items      interface{} `json:"items"`
	Total      int         `json:"total"`
	Page       int         `json:"page"`
	PageSize   int         `json:"pageSize"`
	TotalPages int         `json:"totalPages"`
}

// ParsePagination extracts page and pageSize from query params with sane defaults.
func ParsePagination(r *http.Request, defaultPageSize int) Pagination {
	page := 1
	pageSize := defaultPageSize
	if pageSize <= 0 {
		pageSize = 25
	}

	if p := r.URL.Query().Get("page"); p != "" {
		if v, err := strconv.Atoi(p); err == nil && v > 0 {
			page = v
		}
	}
	if ps := r.URL.Query().Get("pageSize"); ps != "" {
		if v, err := strconv.Atoi(ps); err == nil && v > 0 && v <= 100 {
			pageSize = v
		}
	}

	offset := (page - 1) * pageSize
	return Pagination{Page: page, PageSize: pageSize, Offset: offset}
}

// ApplyPaginationSlice applies pagination to a generic slice by index range.
// Returns (start, end) indices safe for slicing.
func ApplyPaginationSlice(totalLen int, p Pagination) (int, int) {
	start := p.Offset
	if start > totalLen {
		start = totalLen
	}
	end := start + p.PageSize
	if end > totalLen {
		end = totalLen
	}
	return start, end
}

// WritePaginatedJSON writes a standard paginated JSON response.
func WritePaginatedJSON(w http.ResponseWriter, items interface{}, total int, p Pagination) {
	totalPages := total / p.PageSize
	if total%p.PageSize > 0 {
		totalPages++
	}
	resp := PaginatedResponse{
		Items:      items,
		Total:      total,
		Page:       p.Page,
		PageSize:   p.PageSize,
		TotalPages: totalPages,
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resp)
}

File: services/middleware-rs/src/pagination.rs
/// Reusable pagination utilities for 54Bank Rust microservices.
///
/// # Usage
/// ```rust
/// use middleware_rs::pagination::{PaginationParams, paginate_slice};
///
/// let params = PaginationParams::from_query("page=2&pageSize=10");
/// let (items, total_pages) = paginate_slice(&all_items, &params);
/// ```

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct PaginationParams {
    #[serde(default = "default_page")]
    pub page: usize,
    #[serde(default = "default_page_size", rename = "pageSize")]
    pub page_size: usize,
}

fn default_page() -> usize { 1 }
fn default_page_size() -> usize { 25 }

impl PaginationParams {
    pub fn from_query(query: &str) -> Self {
        let mut page = 1usize;
        let mut page_size = 25usize;
        for pair in query.split('&') {
            let mut kv = pair.splitn(2, '=');
            if let (Some(k), Some(v)) = (kv.next(), kv.next()) {
                match k {
                    "page" => { if let Ok(p) = v.parse::<usize>() { if p > 0 { page = p; } } },
                    "pageSize" => { if let Ok(ps) = v.parse::<usize>() { if ps > 0 && ps <= 100 { page_size = ps; } } },
                    _ => {}
                }
            }
        }
        Self { page, page_size }
    }

    pub fn offset(&self) -> usize {
        (self.page - 1) * self.page_size
    }

    pub fn total_pages(&self, total: usize) -> usize {
        if total == 0 { return 0; }
        (total + self.page_size - 1) / self.page_size
    }
}

/// Paginate a slice, returning the sub-slice and total pages.
pub fn paginate_slice<'a, T>(items: &'a [T], params: &PaginationParams) -> (&'a [T], usize) {
    let total = items.len();
    let start = params.offset().min(total);
    let end = (start + params.page_size).min(total);
    let total_pages = params.total_pages(total);
    (&items[start..end], total_pages)
}

#[derive(Serialize)]
pub struct PaginatedResponse<'a, T: Serialize> {
    pub items: &'a [T],
    pub total: usize,
    pub page: usize,
    #[serde(rename = "pageSize")]
    pub page_size: usize,
    #[serde(rename = "totalPages")]
    pub total_pages: usize,
}
